core/src/main/java/com/linecorp/armeria/server/file/AbstractBlockingHttpVfs.java

```diff
@@ -0,0 +1,145 @@
+     * from the specified {@code fileReadExecutor}.
+     */
+    @Override
+    public final HttpFile get(
+            Executor fileReadExecutor, String path, Clock clock,
+            @Nullable String contentEncoding, HttpHeaders additionalHeaders) {
+
+        requireNonNull(fileReadExecutor, "fileReadExecutor");
+        requireNonNull(path, "path");
+        requireNonNull(clock, "clock");
+        requireNonNull(additionalHeaders, "additionalHeaders");
+
+        return HttpFile.from(CompletableFuture.supplyAsync(
+                () -> blockingGet(fileReadExecutor, path, clock, contentEncoding, additionalHeaders),
+                fileReadExecutor));
+    }
+
+    /**
+     * Finds the file at the specified {@code path}.
+     *
+     * @param fileReadExecutor the {@link Executor} which will perform the read operations against the file
+     * @param path an absolute path that starts with {@code '/'}, whose component separator is {@code '/'}
+     * @param clock the {@link Clock} which provides the current date and time
+     * @param contentEncoding the desired {@code 'content-encoding'} header value of the file.
+     *                        {@code null} to omit the header.
+     * @param additionalHeaders the additional HTTP headers to add to the returned {@link HttpFile}.
+     *
+     * @return the {@link HttpFile} at the specified {@code path}
+     */
+    protected abstract HttpFile blockingGet(Executor fileReadExecutor, String path, Clock clock,
+                                            @Nullable String contentEncoding, HttpHeaders additionalHeaders);
+
+    /**
+     * {@inheritDoc} This method invokes {@link #blockingCanList(Executor, String)} from the specified
+     * {@code fileReadExecutor}.
+     */
+    @Override
+    public final CompletableFuture<Boolean> canList(Executor fileReadExecutor, String path) {
+        requireNonNull(fileReadExecutor, "fileReadExecutor");
+        requireNonNull(path, "path");
+
+        if (directoryListingSupported) {
+            return CompletableFuture.supplyAsync(() -> blockingCanList(fileReadExecutor, path),
+                                                 fileReadExecutor);
+        } else {
+            return super.canList(fileReadExecutor, path);
+        }
+    }
+
+    /**
+     * Returns whether the file at the specified {@code path} is a listable directory. This method returns
+     * {@code false} by default.
+     *
+     * @param fileReadExecutor the {@link Executor} which will perform the read operations against the file
+     * @param path an absolute path that starts with {@code '/'}, whose component separator is {@code '/'}
+     * @return {@code true} if the file is a listable directory. {@code false} if the directory does not exist
+     *         or the file listing is not available.
+     */
+    protected boolean blockingCanList(Executor fileReadExecutor, String path) {
+        return false;
```

Comment:
Isn't this `directoryListingSupported`?

Reply:
It has to return `true` only when the path exists and it is a directory. We can't return `true` here because only a subclass has such information.